### PR TITLE
add rpc method getblockheader

### PR DESCRIPTION
**Pull request by mikedennis**
Add the rpc method "getblockheader" with supporting integration unit test from NBitcoin.  Test CanGetBlockFromRPC() assertion is currently failing but not sure what I'm doing wrong - need another set of eyes to see what I missing.

### CODE REVIEW

**Comment by NicolasDorier** [comment]
Check bitcoin code directly, this is more reliable than the doc.

https://github.com/bitcoin/bitcoin/blob/041dad94b047313a17edb0324742cf80cfd550f1/src/rpc/blockchain.cpp#L629

**Comment by dangershony** [comment]
From the core reference just use the time property is fine
https://github.com/bitcoin/bitcoin/blob/041dad94b047313a17edb0324742cf80cfd550f1/src/rpc/blockchain.cpp#L91

**Comment by dangershony** [comment]
@mikedennis whats the progress on this?

**Comment by bokobza** [comment]
Hi @mikedennis, is this ready for review? apart from the median time, there are a few items in TODO.

**Comment by mikedennis** [comment]
@dangershony @bokobza Still needs some work and clean up but I still haven't figured out why the unit test is not passing. I get a different hash back than the genesis. If anyone knows what I'm doing wrong and can point me in the right direction, I'll finish this baby off!

**Comment by bokobza** [comment]
You're not getting the hash you're expecting because the hash is calculated from the BlockHeader's "value" itself.
If you look at Block.cs line 177 you'll see
`h = Hashes.Hash256(this.ToBytes());`
`this` being a BlockHeader.
What I suggest is for you to change this test to Assert all the fields in the BlockHeader, add what's missing in the BlockHeaderModel and leave the hash for last.
Does that make sense?

**Comment by mikedennis** [comment]
@bokobza Ah yes that makes sense. It's because I haven't added all the fields yet. Sounds good! Thanks for the help.

**Comment by bokobza** [comment]
Nice one, I'll have a look soon.

File: Stratis.Bitcoin/Features/RPC/Models/BlockHeaderModel.cs [hunk 1]
```diff
@@ -0,0 +1,84 @@
+﻿using NBitcoin;
+using NBitcoin.DataEncoders;
+
+namespace Stratis.Bitcoin.Features.RPC.Models
+{
+#pragma warning disable IDE1006 // Naming Styles (ignore lowercase)
```
Comment (bokobza) [inline Stratis.Bitcoin/Features/RPC/Models/BlockHeaderModel.cs]: please remove this and look at my comment below on how to deal with these.

File: Stratis.Bitcoin/Features/RPC/Models/BlockHeaderModel.cs [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+                this.merkleroot = blockHeader.HashMerkleRoot.ToString();
+                this.time = blockHeader.Time;
+                byte[] bytes = this.GetBytes(blockHeader.Bits.ToCompact());
+                string encodedBytes = Encoders.Hex.EncodeData(bytes);
+                this.bits = encodedBytes;
+                this.nonce = (int)blockHeader.Nonce;
+            }
+        }
+
+        /// <summary>
```
Comment (bokobza) [inline Stratis.Bitcoin/Features/RPC/Models/BlockHeaderModel.cs]: I know you're basing yourself on what was done previously but here's a better way:
please call the property **V**ersion (Pascal casing) and use the following for it to be serialized/deserialized with a lower-case 
 [JsonProperty(PropertyName = "version")]

File: Stratis.Bitcoin/Features/RPC/Models/BlockHeaderModel.cs [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        /// </summary>
+        public uint version { get; set; }
+
+        /// <summary>
+        /// The merkle root for this block encoded as hex in RPC byte order
+        /// </summary>
+        public string merkleroot { get; set; }
+
+        /// <summary>
+        /// The nonce thich was successful at turning this particular block
```
Comment (bokobza) [inline Stratis.Bitcoin/Features/RPC/Models/BlockHeaderModel.cs]: which*

File: Stratis.Bitcoin/Features/RPC/Controllers/FullNodeController.cs [hunk 4]
```diff
@@ -110,6 +117,28 @@ public GetInfoModel GetInfo()
             return model;
         }
 
+        /// <summary>
+        /// Implements getblockheader RPC call
+        /// </summary>
+        /// <param name="hash">hash of block</param>
```
Comment (bokobza) [inline Stratis.Bitcoin/Features/RPC/Controllers/FullNodeController.cs]: Please use an upper case at the beginning of XML documentation phrases, and a dot in the end.

File: Stratis.Bitcoin/Features/RPC/Controllers/FullNodeController.cs [hunk 5]
```diff
@@ -110,6 +117,28 @@ public GetInfoModel GetInfo()
...
+        /// <param name="isJsonFormat">indicates whether to provide data in Json or binary format</param>
+        /// <returns>The block header rpc format</returns>
+        [ActionName("getblockheader")]
+        public BlockHeaderModel GetBlockHeader(string hash, bool isJsonFormat = true)
+        {
+            Guard.NotNull(this.Chain, nameof(this.Chain));
```
Comment (bokobza) [inline Stratis.Bitcoin/Features/RPC/Controllers/FullNodeController.cs]: The Guard method is only really used to check parameters.
So here you could have Guard.NotEmpty(hash, nameof(hash));
To check that the chain is not null, a simple if(this.chain == null) is sufficient (or != null depending on how you want to do this).

File: Stratis.Bitcoin/Features/RPC/Models/BlockHeaderModel.cs [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    /// <see cref="https://bitcoin.org/en/developer-reference#getblockheader"/>
+    /// </summary>
+    public class BlockHeaderModel
+    {
+        /// <summary>
+        /// Constructs a RPC BlockHeaderModel from a block header object
+        /// </summary>
+        /// <param name="blockHeader">the block header</param>
+        public BlockHeaderModel(BlockHeader blockHeader)
+        {
```
Comment (bokobza) [inline Stratis.Bitcoin/Features/RPC/Models/BlockHeaderModel.cs]: what happens if the block header is null here?

File: Stratis.Bitcoin/Features/RPC/Models/BlockHeaderModel.cs [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+        public string merkleroot { get; set; }
+
+        /// <summary>
+        /// The nonce thich was successful at turning this particular block
+        /// into one that could be added to the best block chain
+        /// </summary>
+        public int nonce { get; set; }
+
+        /// <summary>
+        /// The target threshhold this blocks header had to pass
```
Comment (bokobza) [inline Stratis.Bitcoin/Features/RPC/Models/BlockHeaderModel.cs]: - 1 'h' in threshold
- block's

File: Stratis.Bitcoin/Features/RPC/Models/BlockHeaderModel.cs [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+        /// <param name="blockHeader">the block header</param>
+        public BlockHeaderModel(BlockHeader blockHeader)
+        {           
+            if (blockHeader != null)
+            {
+                this.version = (uint)blockHeader.Version;
+                this.previousblockhash = blockHeader.HashPrevBlock.ToString();
+                this.merkleroot = blockHeader.HashMerkleRoot.ToString();
+                this.time = blockHeader.Time;
+                byte[] bytes = this.GetBytes(blockHeader.Bits.ToCompact());
```
Comment (bokobza) [inline Stratis.Bitcoin/Features/RPC/Models/BlockHeaderModel.cs]: Here it seems you're changing Bits to uint and then back to bytes. Can't you just do Bits.ToBytes()?

File: Stratis.Bitcoin/Features/RPC/Models/BlockHeaderModel.cs [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+                string encodedBytes = Encoders.Hex.EncodeData(bytes);
```
Comment (bokobza) [inline Stratis.Bitcoin/Features/RPC/Models/BlockHeaderModel.cs]: Are you sure this is according to the bitcoin spec? I'm not sure myself, not a tricky question :-)
https://github.com/bitcoin/bitcoin/blob/041dad94b047313a17edb0324742cf80cfd550f1/src/rpc/blockchain.cpp#L650
https://github.com/bitcoin/bitcoin/blob/041dad94b047313a17edb0324742cf80cfd550f1/src/rpc/blockchain.cpp#L94

**Comment by mikedennis** [comment]
@bokobza Thanks for the feedback. I think I've got the issues sorted out in my latest commit. 

Regarding the questions around BlockHeader.Bits serialization. Another alternative would to make this https://github.com/MetacoSA/NBitcoin/blob/9843ab333f93394d70287bec3f3754610abe7897/NBitcoin/Target.cs#L28 NBitcoin method for GetBytes() public and then I could use it directly.  It was a very simple method so I just copied it into the Model. There is definitely some funkiness in the way Bitcoin is serializing this but I believe I am now doing it correctly as the hashes do match up now. Let me know if you see a better way or want me to modify access to the NBitcoin method  instead.

**Comment by dangershony** [comment]
Can to you not just do a normal set
this.Bits = blockHeader.Bits

File: Stratis.Bitcoin.IntegrationTests/RPCTEsts.cs [hunk 10]
```diff
@@ -54,5 +54,50 @@ public void CheckRPCFailures()
...
+                uint256 hash = rpc.GetBlockHash(0);
+                BlockHeader expectedHeader = node.FullNode.Chain?.GetBlock(hash)?.Header;
+                BlockHeader actualHeader = rpc.GetBlockHeader(0);
+
+                // Assert block header fields match
+                Assert.Equal(expectedHeader.Version, actualHeader.Version);
+                Assert.Equal(expectedHeader.HashPrevBlock, actualHeader.HashPrevBlock);
+                Assert.Equal(expectedHeader.HashMerkleRoot, actualHeader.HashMerkleRoot);
+                Assert.Equal(expectedHeader.Time, actualHeader.Time);
+                Assert.Equal(expectedHeader.Bits.ToCompact(), actualHeader.Bits.ToCompact());
```
Comment (bokobza) [inline Stratis.Bitcoin.IntegrationTests/RPCTEsts.cs]: Are expectedHeader.Bits != actualHeader.Bits? I mean do you need to do .ToCompact to verify equality?

File: Stratis.Bitcoin/Features/RPC/Models/BlockHeaderModel.cs [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+    {
+        /// <summary>
+        /// Constructs a RPC BlockHeaderModel from a block header object.
+        /// </summary>
+        /// <param name="blockHeader">The block header.</param>
+        public BlockHeaderModel(BlockHeader blockHeader)
+        {
+            Guard.NotNull(blockHeader, nameof(blockHeader));
+
+            if (blockHeader != null)
```
Comment (bokobza) [inline Stratis.Bitcoin/Features/RPC/Models/BlockHeaderModel.cs]: you can remove this check now that you have the Guard :-)

File: Stratis.Bitcoin/Features/RPC/Models/BlockHeaderModel.cs [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+                this.Bits = encodedBytes;
+                this.Nonce = (int)blockHeader.Nonce;
+            }
+        }
+
+        /// <summary>
+        /// The blocks version number.
+        /// </summary>
+        [JsonProperty(PropertyName = "version")]
+        public uint Version { get; private set; }
```
Comment (bokobza) [inline Stratis.Bitcoin/Features/RPC/Models/BlockHeaderModel.cs]: Lovely, thank you.

File: Stratis.Bitcoin/Features/RPC/Models/BlockHeaderModel.cs [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+        public uint Time { get; private set; }
+
+        /// <summary>
+        /// Convert compact of miner challenge to byte format,
+        /// serialized for transmission via RPC.
+        /// </summary>
+        /// <param name="compact">Compact representation of challenge.</param>
+        /// <returns>Byte representation of challenge.</returns>
+        /// <seealso cref="Target"/>
+        private byte[] GetBytes(uint compact)
```
Comment (bokobza) [inline Stratis.Bitcoin/Features/RPC/Models/BlockHeaderModel.cs]: Ok let's use this method 👍

**Comment by mikedennis** [comment]
@dangershony I wish it was that simple! The Bits.ToString() gives you a uint256 encoded string but the Target constructor needs a "compact" encoded string. See here https://github.com/MetacoSA/NBitcoin/blob/9843ab333f93394d70287bec3f3754610abe7897/NBitcoin/Target.cs#L43

**Comment by NicolasDorier** [comment]
Do I need to change something in NBitcoin to make it more easy?

**Comment by bokobza** [comment]
Lovely, thank you @mikedennis.
Thanks @NicolasDorier I think we're OK for now :-)

**Pull request merged**